src/safe/SafeMessages.ts

```diff
@@ -0,0 +1,117 @@
+import { TypedDataEncoder, TypedDataDomain } from "ethers";
+import Safe, { buildSignatureBytes, Eip1193Provider } from "@safe-global/protocol-kit";
+import SafeApiKit from "@safe-global/api-kit";
+import type { EIP712TypedData } from "@safe-global/types-kit";
+
+import { addressesByNetwork } from "../constants";
+import { ChainId } from "../types";
+import { DOMAIN_NAME, DOMAIN_VERSION } from "../constants/eip712";
+
+export class SafeMessages {
+  private apiKit: SafeApiKit;
+
+  /**
+   * @param address - The address of the Safe
+   * @param chainId - The chainId where the Safe is deployed
+   * @param provider - An EIP1193 compatible provider
+   * @param safeApiKit - A pre-initialized Safe API Kit e.g. when on a network with a 3rd party transaction service deployment
+   */
+  constructor(
+    private address: `0x${string}`,
+    private chainId: ChainId,
+    private provider: Eip1193Provider,
+    safeApiKit?: SafeApiKit
+  ) {
+    this.apiKit = safeApiKit || new SafeApiKit({ chainId: BigInt(chainId) });
+  }
+
+  async signAndSubmit(
+    values: Record<string, unknown>,
+    types: Record<string, Array<{ name: string; type: string }>>,
+    primaryType: string
+  ): Promise<string> {
+    try {
+      const { messageHash } = await this.initiateSigning({
+        types: {
+          ...types,
+        },
+        primaryType,
+        message: values,
+      });
+      return messageHash;
+    } catch (error) {
+      console.error("[signAndSubmit] error", error instanceof Error ? error.message : error);
+      throw error instanceof Error ? error : new Error("Error signing and submitting message");
+    }
+  }
+
+  private async initiateSigning(
+    config: Omit<EIP712TypedData, "domain">
+  ): Promise<{
+    messageHash: `0x${string}`;
+  }> {
+    if (!this.chainId) throw new Error("No chainId found");
+
+    const safe = await Safe.init({
+      provider: this.provider,
+      safeAddress: this.address,
+    });
+
+    const domain = {
```

Comment:
There is `getTypedDataDomain()`

Reply:
It's not worth fixing this as the types are incompatible and circumventing it introduces at least as much code as having this duplicated here. The slight inconsistencies between the typed data domain types in Safe and ethers are just a big ol' mess.

Reply:
Agreed that it's not worth the effort. Resolving.